src/Hazelcast.Net/CP/FencedLock.cs

```diff
@@ -0,0 +1,398 @@
+        public async Task<bool> TryLockAsync()
+        {
+            long fence = await TryLockAndGetFenceAsync(TimeSpan.FromMilliseconds(0)).CfAwait();
+            return fence != InvalidFence;
+        }
+
+        /// <inheritdoc/> 
+        public async Task UnlockAsync()
+        {
+            long threadId = ContextId;
+            long sessionId = _subsystemSession.GetSessionId(CPGroupId);
+
+            VerifyNoLockOnThread(threadId, sessionId, false);
+
+            if (sessionId == CPSubsystemSession.NoSessionId)
+            {
+                _lockedThreadToSession.TryRemove(threadId, out var _);
+                throw new SynchronizationLockException();
+            }
+
+            try
+            {
+                Guid invocationId = Guid.NewGuid();
+                bool stillLockedByCurrentThread = await RequestUnlockAsync(sessionId, threadId, invocationId).CfAwait();
+
+                if (stillLockedByCurrentThread)
+                    _lockedThreadToSession[threadId] = sessionId;
+                else
+                    _lockedThreadToSession.TryRemove(threadId, out var _);
+
+                _subsystemSession.ReleaseSession(CPGroupId, sessionId);
+            }
+            catch (RemoteException e)
+            {
+                if (e is RemoteException { Error: RemoteError.SessionExpiredException })
+                {
+                    _subsystemSession.InvalidateSession(CPGroupId, sessionId);
+                    _lockedThreadToSession.TryRemove(threadId, out var _);
+                    throw new SessionExpiredException(e);
+                }
+                else if (e is RemoteException { Error: RemoteError.IllegalMonitorState })
+                {
+                    _lockedThreadToSession.TryRemove(threadId, out var _);
+                    throw new SynchronizationLockException("IllegalMonitorState exception is thrown by server. Are you trying to unlock without holding?", e);
+                }
+            }
+            catch (Exception)
+            {
+                throw;
+            }
+        }
+        #endregion
+
+        /// <summary>
+        /// Verifies there is no lock between ThreadId->SessionId. Otherwise throws.
+        /// </summary>
+        /// <param name="threadId"></param>
+        /// <param name="sessionId"></param>
+        /// <param name="releaseSession"></param>
+        /// <exception cref="LockOwnershipLostException"></exception>
+        private void VerifyNoLockOnThread(long threadId, long sessionId, bool releaseSession = true)
+        {
+            if (_lockedThreadToSession.TryGetValue(threadId, out var lockedSessionId) && lockedSessionId != sessionId)
+            {
+                _lockedThreadToSession.TryRemove(threadId, out var _);
+
+                if (releaseSession)
+                    _subsystemSession.ReleaseSession(CPGroupId, sessionId);
+
+                throw new LockOwnershipLostException(lockedSessionId.ToString("D"));
```

Comment:
Can you match the message in Java? 
`"Current thread is not owner of the Lock[" + proxyName + "] because its Session["
                + sessionId + "] is closed by server!"`

Reply:
Fixed at https://github.com/hazelcast/hazelcast-csharp-client/pull/590/commits/615acd74e54cd18425eb94dfeb577ca5512e2a61